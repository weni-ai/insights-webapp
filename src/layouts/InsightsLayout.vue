<template>
  <section class="insights-layout">
    <ConnectSidebar class="insights-layout__sidebar" />
    <ConnectTopbar class="insights-layout__topbar" />

    <section class="insights-layout__container">
      <InsightsLayoutHeader />
      <section
        class="insights-layout__insights"
        ref="insightsContent"
      >
        <main
          class="insights__main"
          :style="{
            height: `${mainHeight}vh`,
          }"
        >
          <slot />
        </main>
      </section>
      <ResizableBar />
    </section>
  </section>
</template>

<script>
import { mapState } from 'vuex';

import { pxToVh } from '@/utils/css';

import ResizableBar from '@/components/insights/ResizableBar/index.vue';
import ConnectSidebar from '@/components/connect/ConnectSidebar.vue';
import ConnectTopbar from '@/components/connect/ConnectTopbar.vue';

import InsightsLayoutHeader from '@/components/insights/Layout/Header.vue';

export default {
  name: 'InsightsLayout',

  components: {
    InsightsLayoutHeader,
    ResizableBar,
    ConnectSidebar,
    ConnectTopbar,
  },

  computed: {
    ...mapState({
      contentHeight: (state) => state.resizableBar.contentHeight,
      barHandlerHeight: (state) => state.resizableBar.barHandlerHeight,
    }),

    mainHeight() {
      return (
        pxToVh(this.$refs.insightsContent?.clientHeight) -
          this.contentHeight -
          pxToVh(this.barHandlerHeight) || 0
      );
    },
  },
};
</script>

<style lang="scss" scoped>
$topbarHeight: 88px;

.insights-layout {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-template-rows: auto 1fr;
  grid-template-areas:
    'sidebar topbar'
    'sidebar insights';

  height: 100vh;
  max-height: 100vh;
  width: 100vw;
  max-width: 100vw;

  overflow: hidden;

  padding: 0;

  &__sidebar {
    grid-area: sidebar;
  }

  &__topbar {
    grid-area: topbar;
  }

  &__container {
    position: relative;

    overflow: hidden;

    padding: $unnnic-spacing-sm;

    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-sm;
  }

  &__insights {
    height: calc(100vh - $topbarHeight);

    position: relative;

    grid-area: insights;
    display: grid;
    grid-template-rows: minmax(4fr, 6fr) auto;
    grid-template-columns: auto;

    .insights__main {
      display: flex;
      flex-direction: column;

      background-color: $unnnic-color-neutral-white;

      overflow-y: auto;
      overflow-x: hidden;

      transition: height ease-in-out 0.3s;
    }
  }
}
</style>
