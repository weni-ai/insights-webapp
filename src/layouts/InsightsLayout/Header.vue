<template>
  <header class="insights-layout-header">
    <UnnnicBreadcrumb
      v-if="!isInHome"
      :crumbs="breadcrumbs"
      @crumbClick="handlePath($event.path)"
    />
    <h1 class="header__title">
      {{ isInHome ? 'Insights' : selectedDashboardLabel }}
    </h1>
    <section class="header__filters">
      <UnnnicSelectSmart
        v-model="selectedDashboard"
        :options="dashboards"
        orderedByIndex
      />
      <UnnnicInputDatePicker
        class="filters__date-picker"
        v-model="filterDate"
        size="sm"
      />
    </section>
  </header>
</template>

<script>
import moment from 'moment';

export default {
  name: 'InsightsLayoutHeader',

  data() {
    return {
      dashboards: [],
      selectedDashboard: [],
      filterDate: {
        start: moment().subtract(1, 'day').format('YYYY-MM-DD'),
        end: moment().format('YYYY-MM-DD'),
      },
    };
  },

  created() {
    this.dashboards = [
      {
        value: 'dashboards',
        label: 'Dashboards',
        crumbPath: '/',
        crumbName: 'InÃ­cio',
      },
      { value: 'human-service', label: 'Atendimento Humano' },
    ];
  },

  computed: {
    isInHome() {
      return this.selectedDashboard[0]?.value === 'dashboards';
    },

    breadcrumbs() {
      return this.dashboards.map(({ value, label, crumbPath, crumbName }) => ({
        path: crumbPath || `/${value}`,
        name: crumbName || label,
      }));
    },

    selectedDashboardValue() {
      return this.selectedDashboard[0]?.value;
    },
    selectedDashboardLabel() {
      return this.selectedDashboard[0]?.label;
    },
  },

  methods: {
    handlePath(path) {
      this.$router.push(path);
    },
  },

  watch: {
    selectedDashboard(newSelectedDashboard, oldSelectedDashboard) {
      if (oldSelectedDashboard[0]?.value) {
        this.handlePath(`/${this.selectedDashboardValue}`);
      }
    },
    $route(newRoute) {
      const { path } = newRoute;

      const dashboardRelativeToPath = this.dashboards.find(
        ({ value, crumbPath }) => {
          const isHomePath = path === '/' && [value, crumbPath].includes('/');
          const isValidPath = [value, crumbPath].includes(
            path.replace('/', ''),
          );

          return isHomePath || isValidPath;
        },
      );

      this.selectedDashboard[0] = dashboardRelativeToPath;
    },
  },
};
</script>

<style lang="scss" scoped>
.insights-layout-header {
  display: grid;
  gap: $unnnic-spacing-sm;

  .header__title {
    color: $unnnic-color-neutral-darkest;
    font-size: $unnnic-font-size-title-md;
    font-weight: $unnnic-font-weight-bold;
    font-family: $unnnic-font-family-primary;
    line-height: $unnnic-line-height-large * 2;
  }
  .header__filters {
    display: flex;
    flex-direction: row;
    gap: $unnnic-spacing-xs;

    :deep(.unnnic-select-smart) {
      .dropdown-data {
        z-index: 2;
      }
    }

    .filters__date-picker {
      display: grid;

      :deep(.unnnic-form-input) {
        height: 100%;

        .unnnic-icon {
          // It was necessary to follow bad practices here (px) because
          // of how the component was initially implemented.
          top: $unnnic-spacing-sm - 1px;
        }

        .input {
          font-size: $unnnic-font-size-body-gt;
        }
      }
    }
  }
}
</style>
