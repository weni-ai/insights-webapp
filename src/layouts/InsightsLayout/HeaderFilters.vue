<template>
  <section class="insights-layout-header-filters">
    <UnnnicInputDatePicker
      class="filters__date-picker"
      v-model="filterDate"
      size="sm"
      inputFormat="DD/MM/YYYY"
    />
  </section>
</template>

<script>
import moment from 'moment';

export default {
  name: 'InsightsLayoutHeaderFilters',

  data() {
    return {
      filterDate: {
        start: moment().subtract(1, 'day').format('YYYY-MM-DD'),
        end: moment().format('YYYY-MM-DD'),
      },
    };
  },

  methods: {
    routeUpdateFilterDate() {
      const { query } = this.$route;

      if (query.startDate && query.endDate) {
        this.filterDate = {
          start: query.startDate,
          end: query.endDate,
        };
      }
    },

    retainRouteQueries(newRoute, oldRoute) {
      const oldQueryKeys = Object.keys(oldRoute?.query);

      if (oldQueryKeys.length) {
        this.$router.replace({ name: newRoute.name, query: oldRoute.query });
      }
    },
  },

  watch: {
    $route(newRoute, oldRoute) {
      if (newRoute.name !== oldRoute.name) {
        this.retainRouteQueries(newRoute, oldRoute);
        this.routeUpdateFilterDate();
      }
    },
    filterDate(newDate) {
      this.$router.replace({
        name: this.$route.name,
        query: { startDate: newDate.start, endDate: newDate.end },
      });
    },
  },
};
</script>

<style lang="scss" scoped>
.insights-layout-header-filters {
  display: flex;
  flex-direction: row;
  gap: $unnnic-spacing-xs;

  .filters__date-picker {
    display: grid;

    :deep(.unnnic-form-input) {
      height: 100%;

      .unnnic-icon {
        // It was necessary to follow bad practices here (px) because
        // of how the component was initially implemented.
        top: $unnnic-spacing-sm - 1px;
      }

      .input {
        font-size: $unnnic-font-size-body-gt;
      }
    }
  }
}
</style>
