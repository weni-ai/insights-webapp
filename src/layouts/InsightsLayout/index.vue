<template>
  <section class="insights-layout">
    <ConnectSidebar class="insights-layout__sidebar" />
    <ConnectTopbar class="insights-layout__topbar" />

    <section class="insights-layout__insights">
      <main class="insights__main">
        <slot />
      </main>
      <ResizableBar />
    </section>
  </section>
</template>

<script>
import ResizableBar from '@/components/ResizableBar.vue';
import ConnectSidebar from '@/components/connect/ConnectSidebar.vue';
import ConnectTopbar from '@/components/connect/ConnectTopbar.vue';

export default {
  name: 'InsightsLayout',
  components: { ResizableBar, ConnectSidebar, ConnectTopbar },
};
</script>

<style lang="scss" scoped>
.insights-layout {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-template-rows: auto 1fr;
  grid-template-areas:
    'sidebar topbar'
    'sidebar insights';

  height: 100vh;
  max-height: 100vh;
  width: 100vw;
  max-width: 100vw;

  overflow: hidden;

  padding: 0;

  &__sidebar {
    grid-area: sidebar;
  }

  &__topbar {
    grid-area: topbar;
  }

  &__insights {
    grid-area: insights;

    padding: $unnnic-spacing-ant;

    .insights__main {
      grid-column: 2;

      height: 100%;

      background-color: $unnnic-color-neutral-white;
    }
  }
}
</style>
