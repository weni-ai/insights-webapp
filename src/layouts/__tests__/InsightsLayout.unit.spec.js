import { beforeEach, describe, expect, it, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import InsightsLayout from '@/layouts/InsightsLayout.vue';

describe('InsightsLayout', () => {
  let wrapper;
  let store;
  const mockSetOnboardingRef = vi.fn();

  const createWrapper = () => {
    return mount(InsightsLayout, {
      global: {
        plugins: [store],
        stubs: {
          InsightsLayoutHeader: true,
        },
      },
      slots: {
        default: '<div data-testid="content-slot">Content</div>',
      },
    });
  };

  beforeEach(() => {
    store = createStore({
      modules: {
        dashboards: {
          namespaced: true,
          state: {
            currentDashboardFilters: [],
          },
        },
        onboarding: {
          namespaced: true,
          mutations: {
            SET_ONBOARDING_REF: mockSetOnboardingRef,
          },
        },
      },
    });

    wrapper = createWrapper();
  });

  describe('Initial Render', () => {
    it('renders the layout with InsightsLayoutHeader and default slot', async () => {
      store.state.dashboards.currentDashboardFilters = ['filter1'];
      await wrapper.vm.$nextTick();
      expect(
        wrapper.findComponent({ name: 'InsightsLayoutHeader' }).exists(),
      ).toBe(true);

      const contentSlot = wrapper.find('[data-testid="content-slot"]');
      expect(contentSlot.exists()).toBe(true);
      expect(contentSlot.text()).toBe('Content');
    });

    it('does not display the insights section if there are no currentDashboardFilters', () => {
      const insightsSection = wrapper.find('.insights-layout__insights');
      expect(insightsSection.exists()).toBe(false);
    });
  });

  describe('Dashboard Filters', () => {
    it('displays the insights section when currentDashboardFilters are present', async () => {
      store.state.dashboards.currentDashboardFilters = ['filter1'];
      await wrapper.vm.$nextTick();

      const insightsSection = wrapper.find('.insights-layout__insights');
      expect(insightsSection.exists()).toBe(true);
    });
  });

  describe('Vuex Actions', () => {
    it('calls setOnboardingRef mutation on mount with correct payload', () => {
      expect(mockSetOnboardingRef).toHaveBeenCalledWith(expect.any(Object), {
        key: 'insights-layout',
        ref: wrapper.element,
      });
    });
  });

  it('matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
