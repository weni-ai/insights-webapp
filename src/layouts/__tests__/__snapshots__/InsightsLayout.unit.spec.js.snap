// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InsightsLayout > matches snapshot 1`] = `
<section
  class="insights-layout"
  data-v-abe5f7dc=""
>
  <insights-layout-header-stub
    data-v-abe5f7dc=""
  />
  <!--v-if-->
</section>
`;
